### ISSUE TITLE
[REQ] Cleanup Jackson type info mess

### ISSUE THREAD

**Issue by rpost** [OpenAPITools/openapi-generator#9441]
General question: why some java generators use `JsonTypeInfo.As.EXISTING_PROPERTY` and some `JsonTypeInfo.As.PROPERTY`?

More specific question: why spring-java generator for following specification snippet:

```yml
    Service:
      type: object
      discriminator:
        propertyName: '@type'
      required:
        - '@type'
      properties:
        '@type':
          description: JSON type discriminator, must contain object type name
          type: string
    ConcreteService:
      allOf:
        - $ref: '#/components/schemas/Service'
        - type: object
```

creates sources as follows:

```java
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "@type", visible = true)
@JsonSubTypes({
  @JsonSubTypes.Type(value = ConcreteService.class, name = "ConcreteService"),
})

public class Service   {
  @JsonProperty("@type")
  private String atType;
```

?

Why is discriminator value exposed as object property? It should be invisibile from java developer point of view. What is even worse is that I have to manually set proper type in order to have `@type` filled in serialized json.

Considering yaml above I would expect that

```java
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "@type", visible = true)
@JsonSubTypes({
  @JsonSubTypes.Type(value = ConcreteService.class, name = "ConcreteService"),
})

public class Service   {
// no @type property!
```
is created.

**Comment by marcelstoer** [comment]
I've been there some three years ago: https://stackoverflow.com/questions/50163882/use-swagger-openapi-discriminator-so-that-jackson-serializes-object-correctly, take note of the follow-up issues in the comment.

- swagger-api/swagger-codegen/issues/8137
- FasterXML/jackson-databind/issues/2022

**Comment by rpost** [comment]
I changed my mind about how type discriminators should be configured.

Verba docent, exempla trahunt:

This test summarizes current behaviour for generators which use `EXISTING_PROPERTY`:

```java
package current;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class CurrentBehaviour {

    ObjectMapper mapper = new ObjectMapper();

    @Test
    void testSerialization() throws JsonProcessingException {
        String json = mapper.writeValueAsString(new Child());
        Assertions.assertEquals("{\"@type\":null}", json); //not optimal solution: developer has to set atType manually

        Child child = new Child();
        child.atType = "Child"; //bad! manual and error prone
        json = mapper.writeValueAsString(child);
        Assertions.assertEquals("{\"@type\":\"Child\"}", json);
    }

    @Test
    void testDeserialization() throws JsonProcessingException {
        Child parsed = mapper.readValue("{\"@type\":\"Child\"}", Child.class);
        Assertions.assertNotNull(parsed);
    }

}

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "@type", visible = true)
@JsonSubTypes({
        @JsonSubTypes.Type(value = Child.class, name = "Child"),
})
class Parent {
    @JsonProperty("@type")
    String atType;
}

class Child extends Parent {
}
```

This is my proposal how to improve:

```java
package suggested;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class Suggestion {

    ObjectMapper mapper = new ObjectMapper();

    @Test
    void testSerialization() throws JsonProcessingException {
        String json = mapper.writeValueAsString(new Child());
        Assertions.assertEquals("{\"@type\":\"Child\"}", json); // type discriminator handled by jackson itself!

        Child child = new Child();
        child.atType = "Child"; // even if developers touches this field manually...
        json = mapper.writeValueAsString(child);
        Assertions.assertEquals("{\"@type\":\"Child\"}", json); // ... jackson serializes it only once

        child = new Child();
        child.atType = "nonsense"; // even if developers breaks this ...
        json = mapper.writeValueAsString(child);
        Assertions.assertEquals("{\"@type\":\"Child\"}", json); // ... jackson just ignores error
    }

    @Test
    void testDeserialization() throws JsonProcessingException {
        Child parsed = mapper.readValue("{\"@type\":\"Child\"}", Child.class);
        Assertions.assertNotNull(parsed);
    }

}

@JsonIgnoreProperties(value = "@type", allowSetters = true)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "@type", visible = true)
@JsonSubTypes({
        @JsonSubTypes.Type(value = Child.class, name = "Child"),
})
class Parent {
    @JsonProperty("@type")
    String atType;
}

class Child extends Parent {
}
```

This configures Jackson so that it both:
- fills field in class during deserialization
- automatically fills json property with correct value during serialization

Are you open for MR with such change?

**Comment by rpost** [comment]
Ok, one part solved, we are closer to "good" but still not yet there.

To proceed further we have to clarify what is openapi spec valid and what not.

For sure this is one is valid and should be supported:

```yaml
openapi: 3.0.0
info:
  title: Sample API
  version: 1.0.0
paths: {}
components:
  schemas:
    Dog:
      allOf:
        - $ref: '#/components/schemas/Animal'
        - type: object
          properties:
            breed:
              type: string
    Cat:
      allOf:
        - $ref: '#/components/schemas/Animal'
        - type: object
          properties:
            breed:
              type: string
    Animal:
      type: object
      discriminator:
        propertyName: className
      required:
        - className
      properties:
        className:
          type: string
        color:
          type: string
          default: red
```

But what if `Animal` object doesn't have `className` property:

```yaml
openapi: 3.0.0
info:
  title: Sample API
  version: 1.0.0
paths: {}
components:
  schemas:
    Dog:
      allOf:
        - $ref: '#/components/schemas/Animal'
        - type: object
          properties:
            breed:
              type: string
    Cat:
      allOf:
        - $ref: '#/components/schemas/Animal'
        - type: object
          properties:
            breed:
              type: string
    Animal:
      type: object
      discriminator:
        propertyName: className
      required:
        - className
      properties:
        color:
          type: string
          default: red
```

Is this valid?

Why I'm asking? Seems like some Java generators assume first approach (the ones that had `EXISTING_PROPERTY` before my changes) and some assume second one.

Let's take `jaxrs-cxf-client` for instance.

Given first yaml as input, generated classes look as follow:

```java
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "className", visible = true)
@JsonSubTypes({
  @JsonSubTypes.Type(value = Cat.class, name = "Cat"),
  @JsonSubTypes.Type(value = Dog.class, name = "Dog"),
})
public class Animal  {
  
  @ApiModelProperty(required = true, value = "")
  private String className;

  @ApiModelProperty(value = "")
  private String color = "red";
//...
}
```
and
```java
public class Cat extends Animal {
  
  @ApiModelProperty(value = "")
  private String breed;
//...
}
```

Now, when I try to serialize:

```java
        String s = new ObjectMapper().writeValueAsString(new Cat());
        System.err.println(s);
```

I'm getting `{"className":"Cat","className":null,"color":"red","breed":null}` which is obviously wrong. Interestingly the same generator fed with second yaml works fine (ie: there is no duplication of `className` field in serialized json). Even more interestingly second yaml combined with `spring` generator creates classes which fail to deserialize `{"className":"Cat","className":null,"color":"red","breed":null}` : `Unrecognized field "className" (class de.telekom.hpbxcc.mate.client.model.Cat), not marked as ignorable (2 known properties: "color", "breed"])`

Long story short: to move forward and clean this chaos we need to decide whether second one yaml is valid and supported

**Comment by rpost** [comment]
related: #3796

**Comment by cookeac** [comment]
I'm working with a group of participants on an open source project and dealing with the same issue.
The first yaml that @rpost has listed is our understanding of what should be valid according to the OpenAPI spec, and the second yaml is what many Java developers have needed to implement until now so that openapi-generator would generate the correct serialisation/deserialisation. The problem is that the second yaml does not seem to align with the spec, and does not produce proper code for other languages.
Moving to support the first yaml is the right move - but do you need to also support the second yaml for backwards compatibility for Java-only use?

**Comment by rpost** [comment]
Thank you @cookeac for your opinion. Sad thing is that apparently no one else (from maintainers) is interested in this topic. Pity, I was eager to clean this up.

**Comment by nickshoe** [comment]
I think that we should follow the "valid OpenAPI spec" path and consider these two requirements:
a. we want the serialization process to produce a JSON key indicating the object type, as specified with `@JsonSubTypes` annotation, **without having to specify an additional property in the POJO** since Java objects already have this information represented by their (sub)classes
b. we want the deserialization process to produce POJO instances of the correct subclasses

I'm starting to think that this is rather a problem with how Jackson handles serialization/deserialization itself: the `PROPERTY` strategy seems to be working just fine during serialization, but requires a regular POJO property during deserialization. Anyways, there is a workaround for the deserialization issue: https://github.com/OpenAPITools/openapi-generator/pull/4588#issuecomment-763159780.

**Comment by nickshoe** [comment]
Another proposal may be the following (but I haven't tested yet):
- use `PROPERTY` strategy **and set `visible = false`** [here](https://fasterxml.github.io/jackson-annotations/javadoc/2.4/com/fasterxml/jackson/annotation/JsonTypeInfo.html#visible())
- don't use an additional property in the POJO

By setting `visible = false` the deserialization process should not complain about unrecognized fields/properties. This way, there's no need to change the `ObjectMapper` configuration as I've pointed out in my previous comment.

**Comment by nickshoe** [comment]
Ok, I confirm the proposal made in my latest comment.

I've tested it, here is another comment of mine in another issue: https://github.com/OpenAPITools/openapi-generator/issues/3796#issuecomment-1199994908.

The `PROPERTY` strategy without `visible = true` (its default value is false, thus we can omit `visible = false`) is my preferred solution for this problem.

**Comment by smil2k** [comment]
> Ok, I confirm the proposal made in my latest comment.
> 
> I've tested it, here is another comment of mine in another issue: [#3796 (comment)](https://github.com/OpenAPITools/openapi-generator/issues/3796#issuecomment-1199994908).
> 
> The `PROPERTY` strategy without `visible = true` (its default value is false, thus we can omit `visible = false`) is my preferred solution for this problem.

I tested your proposal, I came to this conclusion also. Visible=false will hide the type property from the object. (property / visible = false)
- No need to ignore unknown stuff
- type is not visible, encoded in the class system

The corrent mode of operation is just wrong:
as=property, visible = serialized by jackson, property is serialized = 2x!

It would be good to have some kind of config option to set which options to use. 
This is up to the taste of the developer.

- **as=existing property / visible  not serialized by jackson, property is serialized = 1x**
- **as=property, not-visible = serialized by jackson + property is not in the class = 1x**
